File: sources/vimm.py
"""Vimm's Lair direct-download source for Gamarr."""
import logging
import re

import requests
import urllib3

urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

from .base import GameSource

logger = logging.getLogger("gamarr.sources.vimm")

_VIMM_SYSTEM_MAP = {
    "nes": "NES", "snes": "SNES", "n64": "N64", "ngc": "GameCube",
    "wii": "Wii", "gb": "GB", "gbc": "GBC", "gba": "GBA", "nds": "DS",
    "genesis": "Genesis", "saturn": "Saturn", "dc": "Dreamcast",
    "psx": "PS1", "ps2": "PS2", "ps3": "PS3", "psp": "PSP",
    "xbox": "Xbox", "xbox360": "Xbox360",
}


class VimmSource(GameSource):
    name = "vimm"
    label = "Vimm's Lair"
    color = "#6366f1"
    source_type = "ddl"

    config_fields = []

    def enabled(self) -> bool:
        return True

    def search(self, query: str, platform_slug: str = None) -> list:
        results = []
        reverse_map = {v: k for k, v in _VIMM_SYSTEM_MAP.items()}

        params = {"p": "list", "q": query}
        if platform_slug and platform_slug in _VIMM_SYSTEM_MAP:
            params["system"] = _VIMM_SYSTEM_MAP[platform_slug]

        system_from_filter = _VIMM_SYSTEM_MAP.get(platform_slug, "")

        try:
            resp = requests.get(
                "https://vimm.net/vault/",
                params=params,
                headers={
                    "User-Agent": (
                        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                        "AppleWebKit/537.36 (KHTML, like Gecko) "
                        "Chrome/120.0.0.0 Safari/537.36"
                    )
                },
                timeout=15,
                verify=False,
            )
            if resp.status_code != 200:
                return []

            # Parse game links: <a href="/vault/{ID}" ...>Game Name</a>
            for match in re.finditer(
                r'<a\s+href=\s*"/vault/(\d+)"[^>]*>([^<]+)</a>',
                resp.text,
            ):
                game_id = match.group(1)
                game_name = match.group(2).strip()

                slug = platform_slug
                system_clean = system_from_filter or "Unknown"
                is_pc = False

                # Detect platform from title when no filter is set, e.g. "Castlevania (NES)"
                if not slug:
                    title_system = re.search(r'\(([A-Za-z0-9]+)\)\s*$', game_name)
                    if title_system:
                        sys_name = title_system.group(1)
                        reverse_map_lower = {v.lower(): k for k, v in _VIMM_SYSTEM_MAP.items()}
                        if sys_name.lower() in reverse_map_lower:
                            slug = reverse_map_lower[sys_name.lower()]
                            system_clean = sys_name

                display_title = game_name
                if system_clean and system_clean != "Unknown" and f"({system_clean})" not in game_name:
                    display_title = f"{game_name} ({system_clean})"

                results.append({
                    "title": display_title,
                    "size": 0,
                    "size_human": "?",
                    "seeders": 0,
                    "leechers": 0,
                    "indexer": "Vimm's Lair",
                    "download_url": "",
                    "magnet_url": "",
                    "info_hash": "",
                    "guid": f"https://vimm.net/vault/{game_id}",
                    "platform": system_clean,
                    "platform_slug": slug,
                    "is_pc": is_pc,
                    "age": 0,
                    "source_type": "ddl",
                    "vimm_id": game_id,
                    "safety_score": 90,
                    "safety_warnings": [],
                })
        except Exception as e:
            logger.warning(f"Vimm search error: {e}")
        return results[:20]

File: internal/search/vimm.go
package search

import (
	"crypto/tls"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"

	"gamarr/internal/models"
	"gamarr/internal/sources"
)

// VimmPlatformSlugs returns all platform slugs Vimm supports per the runtime
// sources registry.
func VimmPlatformSlugs(reg *sources.Registry) []string {
	slugs := make([]string, 0, len(reg.Vimm.PlatformSystems))
	for s := range reg.Vimm.PlatformSystems {
		slugs = append(slugs, s)
	}
	return slugs
}

// vimmGameRe matches a vault game link. Group 1 is the numeric id, group 2 is
// the attribute string, group 3 is the link text.
var vimmGameRe = regexp.MustCompile(`<a\s+href=\s*"/vault/(\d+)"([^>]*)>([^<]+)</a>`)
var vimmFlagRe = regexp.MustCompile(`class="flag" title="([^"]+)"`)
var vimmTDRe = regexp.MustCompile(`(?is)<td[^>]*>(.*?)</td>`)
var vimmTagRe = regexp.MustCompile(`<[^>]+>`)
var vimmSysTokenRe = regexp.MustCompile(`^[A-Za-z][A-Za-z0-9-]{0,24}$`)
var vimmDigitRe = regexp.MustCompile(`^\d+$`)
var vimmHiddenRe = regexp.MustCompile(`(?i)display:\s*none`)
var titleSysRe = regexp.MustCompile(`\(([A-Za-z0-9]+)\)\s*$`)

type vimmHit struct {
	ID, Title, System, Region string
}

func parseVimmSearchHTML(body string) []vimmHit {
	var hits []vimmHit
	seen := make(map[string]bool)
	for _, loc := range vimmGameRe.FindAllStringSubmatchIndex(body, -1) {
		id := body[loc[2]:loc[3]]
		attrs := body[loc[4]:loc[5]]
		title := strings.TrimSpace(body[loc[6]:loc[7]])
		if id == "999999" || vimmDigitRe.MatchString(title) || vimmHiddenRe.MatchString(attrs) {
			continue
		}
		if seen[id] {
			continue
		}
		seen[id] = true
		row := surroundingTR(body, loc[0])
		hits = append(hits, vimmHit{
			ID:     id,
			Title:  title,
			System: systemFromRow(row),
			Region: flagsFrom(row),
		})
	}
	return hits
}

func surroundingTR(body string, pos int) string {
	start := strings.LastIndex(body[:pos], "<tr")
	if start < 0 {
		return ""
	}
	end := strings.Index(body[pos:], "</tr>")
	if end < 0 {
		return body[start:]
	}
	return body[start : pos+end+len("</tr>")]
}

func flagsFrom(row string) string {
	if row == "" {
		return ""
	}
	var regions []string
	seen := make(map[string]bool)
	for _, m := range vimmFlagRe.FindAllStringSubmatch(row, -1) {
		if seen[m[1]] {
			continue
		}
		seen[m[1]] = true
		regions = append(regions, m[1])
	}
	return strings.Join(regions, ", ")
}

func systemFromRow(row string) string {
	if row == "" {
		return ""
	}
	m := vimmTDRe.FindStringSubmatch(row)
	if m == nil {
		return ""
	}
	inner := strings.TrimSpace(m[1])
	if strings.Contains(strings.ToLower(inner), "<a") {
		return ""
	}
	inner = strings.TrimSpace(vimmTagRe.ReplaceAllString(inner, ""))
	if vimmSysTokenRe.MatchString(inner) {
		return inner
	}
	return ""
}

// SearchVimm searches Vimm's Lair for ROMs.
func SearchVimm(reg *sources.Registry, query string, platformSlug string) []*models.SearchResult {
	if IsCircuitOpen("vimm") {
		slog.Warn("vimm circuit open, skipping search")
		return nil
	}
	params := url.Values{"p": {"list"}, "q": {query}}
	if platformSlug != "" {
		if sys, ok := reg.Vimm.PlatformSystems[platformSlug]; ok {
			params.Set("system", sys)
		}
	}
	systemFromFilter := reg.Vimm.PlatformSystems[platformSlug]

	client := &http.Client{
		Timeout: 15 * time.Second,
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
		},
	}
	req, _ := http.NewRequest("GET", reg.Vimm.BaseURL+"?"+params.Encode(), nil)
	req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36")

	resp, err := client.Do(req)
	if err != nil {
		slog.Warn("Vimm search error", "error", err)
		RecordSearchFail("vimm", err.Error())
		return nil
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		RecordSearchFail("vimm", fmt.Sprintf("HTTP %d", resp.StatusCode))
		return nil
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil
	}

	reverseMap := make(map[string]string)
	for slug, sys := range reg.Vimm.PlatformSystems {
		reverseMap[strings.ToLower(sys)] = slug
	}

	var results []*models.SearchResult
	for _, hit := range parseVimmSearchHTML(string(body)) {
		slug := platformSlug
		systemClean := systemFromFilter
		if hit.System != "" {
			systemClean = hit.System
			if s, ok := reverseMap[strings.ToLower(hit.System)]; ok {
				slug = s
			} else if platformSlug == "" {
				slug = ""
			}
		}
		if systemClean == "" {
			systemClean = "Unknown"
		}
		if hit.System == "" && slug == "" {
			if ts := titleSysRe.FindStringSubmatch(hit.Title); ts != nil {
				sysName := ts[1]
				if s, ok := reverseMap[strings.ToLower(sysName)]; ok {
					slug = s
					systemClean = sysName
				}
			}
		}

		displayTitle := hit.Title
		if hit.Region != "" && !strings.Contains(hit.Title, hit.Region) {
			displayTitle = fmt.Sprintf("%s (%s)", displayTitle, hit.Region)
		}
		if systemClean != "" && systemClean != "Unknown" && !strings.Contains(displayTitle, "("+systemClean+")") {
			displayTitle = fmt.Sprintf("%s (%s)", displayTitle, systemClean)
		}

		results = append(results, &models.SearchResult{
			Title:          displayTitle,
			SizeHuman:      "?",
			Indexer:        "Vimm's Lair",
			GUID:           fmt.Sprintf("%s%s", reg.Vimm.BaseURL, hit.ID),
			Platform:       systemClean,
			PlatformSlug:   slug,
			SourceType:     "ddl",
			VimmID:         hit.ID,
			SafetyScore:    90,
			SafetyWarnings: []string{},
		})
	}
	if len(results) > 20 {
		results = results[:20]
	}

	RecordSearchSuccess("vimm")
	return results
}
